Added additional positioning properties to Popup and ContextMenu.

grokys opened the pull request:
## What does the pull request do?

Adds the following positioning properties to `Popup`:

- `PlacementAnchor`
- `PlacementGravity`
- `PlacementRect`
- `PlacementConstraintAdjustment`

Adds the new properties to `ContextMenu`, along with the following which were previously present on `Popup` but not `ContextMenu`:

- `PlacementMode`
- `PlacementTarget`
- `HorizontalOffset`
- `VerticalOffset`

These properties are passed to the inner `Popup` to position the context menu popup.

grokys committed:
Added nullable annotations to ContextMenu.

grokys committed:
Added positioning properties to ContextMenu.

MarchingCube commented on src/Avalonia.Controls/ContextMenu.cs:
```diff
@@ -47,23 +71,60 @@ public ContextMenu(IMenuInteractionHandler interactionHandler)
         }
 
+        /// <summary>
+        /// Gets or sets the placement mode of the popup in relation to the <see cref="PlacementTarget"/>.
```
```suggestion
        /// Gets or sets the placement mode of the context menu in relation to the <see cref="PlacementTarget"/>.
```

Wouldn't this be clearer?

grokys committed:
Added/tweaked docs.

grokys committed:
Added additional placement properties to PopupRoot.

And split `PopupPositioningEdge` into `PopupAnchor`  and `PopupGravity` so their values can be documented more clearly.

grokys committed:
Added additional placement properties to ContextMenu.

grokys commented on src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs:
```diff
@@ -50,46 +50,48 @@
-    /// an axis, or flipping around a rectangle. These positioner-created rules are
-    /// constrained by the requirement that a popup must intersect with or
-    /// be at least partially adjacent to its parent surface.
+    /// Provides positioning parameters to <see cref="IPopupPositioner"/>.
```
Tweaked the XML docs to follow normal .NET conventions (short description in summary, longer description in remarks)

grokys commented on src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs:
```diff
@@ -171,123 +177,249 @@ public enum PopupPositionerConstraintAdjustment
+    /// Defines the edges around an anchor rectangle on which a popup will open.
+    /// </summary>
     [Flags]
-    public enum PopupPositioningEdge
```
@kekekeks I split this enum into two separate enums, because it allows us to more easily explain what the enum values do in the case of anchors and gravity than when it's a single enum. I notice that in the xdg spec it's two enums anyway.

grokys replied:
I know this is a breaking change, but IMO we should be doing breaking changes now if it makes things easier to understand. I had difficulty understanding what "gravity" was, I think this makes it clearer.

grokys commented on src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs:
```diff
@@ -317,32 +449,33 @@ static class PopupPositionerExtensions
-                    .TransformToAABB(matrix.Value);
+                var bounds = new Rect(default, target.Bounds.Size);
+                var anchorRect = rect ?? bounds;
+                positionerParameters.AnchorRectangle = anchorRect.Intersect(bounds).TransformToAABB(matrix.Value);
```
@kekekeks I added an intersection with `bounds` here because from what I understand the anchor rectangle needs to be within the control - is this correct?

kekekeks replied:
We need to check how WPF handles the placement rect, but for me it makes sense to intersect with the control bounds, yes.

kekekeks replied:
Actually, there might be a problem when placement rect is somehow outside of the control area. We might end up with a 0x0 rect which is invalid.

grokys replied:
Well a control's bounds may also be of size 0,0... Maybe we should treat a 0-sized rect as a one-sized rect?

grokys committed:
Tweak doc comments.

MarchingCube commented on src/Avalonia.Controls/Primitives/PopupPositioning/ManagedPopupPositioner.cs:
```diff
@@ -25,6 +25,10 @@ public ManagedPopupPositionerScreenInfo(Rect bounds, Rect workingArea)
 
+    /// <summary>
+    /// An <see cref="IPopupPositioner"/> implementation for platforms on which a popup can be
+    /// aritrarily positioned.
```
```suggestion
    /// arbitrarily positioned.
```

MarchingCube commented on src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs:
```diff
@@ -50,46 +50,48 @@
+    /// Provides positioning parameters to <see cref="IPopupPositioner"/>.
     /// </summary>
+    /// <remarks>
+    /// The IPopupPositioner provides a collection of rules for the placement of a a popup relative
```
```suggestion
    /// The IPopupPositioner provides a collection of rules for the placement of a popup relative
```

MarchingCube commented on src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs:
```diff
@@ -50,46 +50,48 @@
         /// </summary>
-        public PopupPositioningEdge Anchor
+        /// <remarks>
+        /// The specified anchor is used derive an anchor point that the popup will be positioned
```
```suggestion
        /// The specified anchor is used to derive an anchor point that the popup will be positioned
```

grokys committed:
Merge branch 'master' into feature/contentmenu-positioning

grokys committed:
Added PlacementConstraintAdjustment property.

To `Popup` and `ContextMenu`. Set default value to be `FlipX | FlipY` to try to match WPF's defaults.

grokys committed:
Added ResizeX/Y constraint adjustments.

And set these by default on `Popup`/`ContextMenu`.

Commit:
Merge branch 'master' into feature/contentmenu-positioning

jmacato committed:
Merge branch 'master' into feature/contentmenu-positioning

jmacato commented:
yolo merging this
